PEP 737: Unify type name formatting (#3554)
python/peps: peps/pep-0737.rst

place, by the ``type.__repr__()`` implementation. Using
``Py_TYPE(obj)->tp_name`` is more convenient than calling
``PyType_GetQualName()`` which requires ``Py_DECREF()``. Moreover,
``PyType_GetQualName()`` was only added recently, in Python 3.11.

Some functions use ``%R`` to format a type name. Example::

    PyErr_Format(PyExc_TypeError,
                 "calling %R should have returned an instance "
                 "of BaseException, not %R",
                 type, Py_TYPE(value));

Using PyTypeObject.tp_name is inconsistent with Python
------------------------------------------------------

The ``PyTypeObject.tp_name`` member is different depending on the type
implementation:

* Static types and heap types in C: type fully qualified name.
* Python class: type short name (``type.__name__``).

So using ``Py_TYPE(obj)->tp_name`` to format an object type name gives
a different output depending if a type is implemented in C or in Python.

It goes against :pep:`399` "Pure Python/C Accelerator Module
Compatibility Requirements" principles which recommends code to behave
the same way if written in Python or in C.

> Example::

Author: Is `pycon` a typo or a legit programming language? Hmm.

> Example::

Author: Oh nice! I didn't know this one.

  using colon (``:``) separator.
* ``f"{type(obj):T}"`` formats ``type(obj).__name__``.
* ``f"{type(obj):#T}"`` formats ``type(obj).__fully_qualified_name__``.

Using short format (such as ``z``, a single letter) requires to refer to
format documentation to understand how a type name is formatted, whereas
``type(obj).__name__`` is explicit.

The dot character (``.``) is already used for the "precision" in format
strings. The colon character (``:``) is already used to separated the
expression from the format specification. For example, ``f"{3.14:g}"``
uses ``g`` format which comes after the colon (``:``). Usually, a format
type is a single letter, such as ``g`` in ``f"{3.14:g}"``, not ``M.T``
or ``M:T``. Reusing dot and colon characters for a different purpose can
be misleading and make the format parser more complicated.

Add !t formatter to get an object type
--------------------------------------

Use ``f"{obj!t:T}"`` to format ``type(obj).__name__``, similar to
``f"{type(obj).__name__}"``.

When the ``!t`` formatter was proposed in 2018, `Eric Smith was opposed
to this
<https://mail.python.org/archives/list/python-dev@python.org/message/BMIW3FEB77OS7OB3YYUUDUBITPWLRG3U/>`_;

> Eric is the author of :pep:`498` "f-string".

Reviewer: ```suggestion
Eric is the author of the f-string :pep:`498` "Literal String Interpolation".
```

Add formats to str % args
-------------------------

It was proposed to add formats to format a type name in ``str % arg``.
For example, ``%T`` and ``%#T`` formats.

> Nowadays, f-string is preferred for new code.

Reviewer: ```suggestion
Nowadays, f-strings are preferred for new code.
```

In short, it is safe to use ``Py_TYPE(obj)`` borrowed reference while
formatting an error message.

If the ``%T`` format expects an instance, formatting a type cannot use
the ``%T`` format, whereas it's a common operation in stdlib C
extensions. The ``%T`` format would only cover half of cases (only
instances). If the ``%T`` format takes a type, all cases are covered
(types and instances using ``Py_TYPE()``).

Other proposed APIs to get a type fully qualified name
------------------------------------------------------

* ``type.__fullyqualname__`` attribute name: attribute without underscore
  between words. Several dunders, including some of the most recently
  added ones, include an underscore in the word: ``__class_getitem__``,
  ``__release_buffer__``, ``__type_params__``, ``__init_subclass__`` and
  ``__text_signature__``.
* ``type.__fqn__`` attribute name (Fully Qualified Name: FDN).
* Add a function to the ``inspect`` module. Need to import the
  ``inspect`` module to use it.

Omit __main__ module in the type fully qualified name
-----------------------------------------------------

The ``pdb`` module formats a type fully qualified names in a similar way

> than proposed ``type.__fully_qualified_name__``, but it omits the module

Reviewer: ```suggestion
as the proposed ``type.__fully_qualified_name__``, but it omits the module
```

Title: Unify type name formatting
Author: Victor Stinner <vstinner@python.org>
Status: Draft
Type: Standards Track
Created: 29-Nov-2023
Python-Version: 3.13

Abstract
========

Add new convenient APIs to format type names the same way in Python and
in C. No longer format type names differently depending on how types are
implemented. Also, put an end to truncating type names in C. The new C
API is compatible with the limited C API.

Rationale
=========

Standard library
----------------

In the Python standard library, formatting a type name or the type name
of an object is a common operation to format an error message and to
implement a ``__repr__()`` method. There are different ways to format a
type name which give different outputs.

Example with the ``datetime.timedelta`` type:

* The type short name (``type.__name__``) and the type qualified name
  (``type.__qualname__``) are ``'timedelta'``.
* The type module (``type.__module__``) is ``'datetime'``.
* The type fully qualified name is ``'datetime.timedelta'``.

Python code
^^^^^^^^^^^

In Python, ``type.__name__`` gets the type "short name", whereas
``f"{type.__module__}.{type.__qualname__}"`` formats the type "fully

> qualified name. Usually, ``type(obj)`` or ``obj.__class__`` are used to

Author: ```suggestion
qualified name". Usually, ``type(obj)`` or ``obj.__class__`` are used to
```
